### DISCUSSION TITLE
Type hints for result objects from select

### DISCUSSION THREAD

**Discussion by sector119** [sqlalchemy/sqlalchemy#10760] [Usage Questions] [answered]
Hello,

What is the proper way to hint result object? I know that there is no way for SA to expose tuple content types right now.
But to make it more comfortable should we use Protocol or custom abstract model class as return type for function that return select result objects?

Like 

```
class ResultHint(typing.Protocol):
  col1: int
  col2: str
  col3: list[int]
```

and use typing.Sequence[ResultHint] as return type hint?

Thanks

**Comment by zzzeek** [comment, marked as answer]
we dont have any documented way to type the rows with the column names, just as tuples with the individual types like `Row[Tuple[int, str, list[int]]]`.  in 2.1 we will finally fix the tuple part and you'll get `Row[int, str, list[int]]`.   but for the names, sure a protocol seems like the best idea.

**Reply by sector119** [reply]
Thank you, Mike !

**Reply by chen-tamarin** [reply]
Is the ability to type the rows by field names on the roadmap (near or far)?

**Reply by CaselIT** [reply]
Not currently, since there is currently no way of automatically creating a protocol like in the example in a way that would be picked up by type checker.
If you already have the protocol is quite easy to use `cast` to change the type of the returned sequence in the type your prefer

**Reply by liverpool1026** [reply]
How about typing a statement? Will you recommend doing the following?

stmt: sa.Select[ResultHint] = sa.select(Table.col1, Table.col2, Table.col3)

Thanks

**Reply by CaselIT** [reply]
Not sure what result hint is, but in 2.0 you can do something like (let's assume the types of the 3 columns are int, str, int)
```py
stmt: sa.Select[tuple[int, str, int]] = sa.select(Table.col1, Table.col2, Table.col3)
```

in 2.1 the above typing will instead be


```py
stmt: sa.Select[int, str, int] = sa.select(Table.col1, Table.col2, Table.col3)
```

**Reply by liverpool1026** [reply]
ResultHint is the Protocol that OP has defined.

class ResultHint(typing.Protocol):
  col1: int
  col2: str
  col3: list[int]

**Reply by CaselIT** [reply]
Without cast/typeignore I don't think you can do that, since the type checker is likely to complain on assignment.

Other than that it shouldn't cause any issues in sqlalchemy
